### PR TITLE
Sort dated todos

### PR DESCRIPTION
**Pull request by gisikw**
Replaces #4980 

Sorts dated todo tasks, and also doesn't totally break all the things.

(Which is to say, the `removeTask` function has been rewritten to expect an explicit task, rather than a list and index, and references to the function have been modified to reflect the new signature).

### CODE REVIEW

File: website/views/shared/tasks/task.jade [hunk 1]
```diff
@@ -44,7 +44,7 @@ li(bindonce='list', id='task-{{::task.id}}', ng-repeat='task in obj[list.type+"s
           span.glyphicon.glyphicon-bullhorn(tooltip=env.t('challenge'))
           | &nbsp;
       // delete
-      a(ng-if='!task.challenge.id', ng-click='removeTask(obj[list.type+"s"], $index)', tooltip=env.t('delete'))
```
Comment (crookedneighbor) [inline website/views/shared/tasks/task.jade:47]: Does anyone know why this was done this way in the first place? `git blame` says this line was last touched by Marcelo Sizer, if that helps.

Comment (gisikw) [inline website/views/shared/tasks/task.jade:47]: Yep - method was made here: https://github.com/HabitRPG/habitrpg/commit/e45d8307e713bb2b7732beb59567f4fa2cd102f2#diff-7a6d770c287cf03133942144e9f2c670R45

It looks like at one point, the method was used to both delete the task, as well as modify the UI-bound list array. so the list needed to be passed in. That's no longer the case, but the method signature was just never changed.

Comment (crookedneighbor) [inline website/views/shared/tasks/task.jade:47]: Brilliant detective work!

Comment (gisikw) [inline website/views/shared/tasks/task.jade:47]: :mag:

**Comment by crookedneighbor** [comment]
:+1: :+1: :+1: :+1:

**Comment by Alys** [comment]
I can't delete challenge tasks when editing a challenge that I created, using the latest `develop`. 

Probably this needs to change but I haven't looked at it closely:
website/public/js/controllers/challengesCtrl.js:189:    $scope.removeTask = function(list, $index) {

**Comment by Alys** [comment]
Sorry, I mean, I can't delete challenge tasks when editing a challenge that I created.

**Comment by gisikw** [comment]
EWWWWWWWW DUPLICATION!

Dangit. Gimme 5 min.

**Comment by gisikw** [comment]
I ack'd for `removeTask`, and it looks like challengesCtrl is the only other example. I'll get that pushed up. Feel free to either revert or just hold on for a PR inc 5 min.

**Comment by crookedneighbor** [comment]
We can hold on. It's not live yet. I'll try to get some tests written to cover these functions.

**Comment by crookedneighbor** [comment]
Really, I shouldn't have merged without reviewing/writing some tests first. I just got trigger happy.

**Comment by gisikw** [comment]
Revised estimate: 10-20min. With challenges, removeTask actually _does_ need to modify the list (since we can't just delete a user task, as with regular challenges). I need to:
- Update view to call `removeTask(task, list)`
- Leave task controller alone (it'll just ignore the list arg)
- Update challenges controller to splice the list from the task, rather than the id

Shouldn't be too hard :)

**Comment by gisikw** [comment]
Fix in PR #5000

**Comment by Alys** [comment]
Darnit, there's no revert link. This will have to be reverted manually.

But first:

@gisikw We LOVE your work and the vast bulk of it has been awesome -- and this one will be too eventually :) -- but as far as I can tell from testing tonight, it's causing a glitch with sorting other kinds of tasks (possibly only after the dated To-Do tab has been visited). So, I'm reverting this and #5020 (already done) until we can work out what's wrong.

After I've reverted this, I'll be pushing to beta and then testing there to see if reverting this has fixed the problem I noticed. I'll give you another update when it's all done.

**Comment by gisikw** [comment]
You know what? I bet the orderBy null case is forceably re-sorting stuff. If that assumption is correct, you shouldn't need to visit the dated tab to repro the issue. I'll see what I can get whipped up on my local install.

**Comment by Alys** [comment]
Hokay, PR made, Travis build is building, so I have some moments to breathe. :)

@gisikw Are you familiar with the sorting issues you get in To-Dos when you have completed and archived To-Dos? You can move a To-Do to a new spot and immediately it jumps back or other To-Dos jump around? In beta, which has this PR (and others) I was getting the same behaviour for all three types of tasks (probably Rewards too but I didn't test them). The cause of it would have been different of course, but the symptoms were the same.

Basically, load up a test account with about a dozen tasks in each column and drag a few of them up or down and pretty soon you'll see the sort errors.

**Comment by gisikw** [comment]
Yeah, that makes sense. I'm pouring through the Angular source code now - if this issue is what I suspect, it would apply to all task collections. Investigating now (got sidetracked by a workmergency).

**Comment by Alys** [comment]
And importantly, the sorting problems were happening only in Chrome, not in Firefox (confirmed with two users, one of which had two accounts). No idea why. 

Interestingly, when I went to Firefox, my tasks were sorted correctly, when I went to Chrome, they were missorted. Re-sorting in either browser made no different to the sort order in the other browser. Even less idea why that happened? Aggressive local storage?

**Comment by Alys** [comment]
@gisikw Don't feel that this needs to be fixed urgently. It's reverted now and I'm about to test on beta, so real life can take precedence.

**Comment by gisikw** [comment]
Okay, so looking at https://github.com/angular/angular.js/blob/master/src/ng/filter/orderBy.js#L162, it's trying to do an object comparison. My guess is that behavior varies for some reason by browser.

Anyway, the tl;dr is that I did something stupid, and passing a nil orderBy only _seemed_ to work. The appropriate solution here is to create a custom filter (probably something intuitive like `orderByIf(orderer, predicate)` that will simply delegate to `orderBy` if the predicate is true.

Shouldn't be too much work. Sorry about the issues here - I've largely migrated to React for my frontend JS stuff, so my Angular-fu has stagnated :-/

**Comment by Alys** [comment]
Don't worry about it! These things happen. I'm just happy that my dated To-Dos tooltip is back. ;)

**Comment by gisikw** [comment]
Haha! So it was all a setup! :P

**Comment by gisikw** [comment]
screenshot 2015-04-13 06 59 50

**Comment by Alys** [comment]
Yes, I've been recalling that.

I will never say that again.

**Comment by Alys** [comment]
The emoticons are still appropriate though. ;)

**Comment by gisikw** [comment]
Hmmm, I'm able to reproduce the issue when ordering tasks on Todos > Done tab. If I drag and reorder, the order isn't persisted on refresh. I'm not seeing the behavior with other tabs though. 

I should still be able to fix it, but it would be nice to have a reproducible case for other tasks, so I can write a test for it.

To confirm, the error you were seeing was:
- In any of the task lists
- Create a bunch of tasks
- Drag 'em up and down until they're shuffled
- Refresh
- Tasks aren't ordered to match the shuffle

Is that correct?

**Comment by deilann** [comment]
I didn't have to move tasks at all with this issue. On beta, my tasks in
Chrome were in a different order than they were in Firefox.

**Comment by gisikw** [comment]
:-/ Firefox can't load the login page for my local env, which makes that a difficult use case to test :-/

**Comment by Alys** [comment]
@gisikw Not exactly as you list. It's like this:
- In any of the task lists
- Create a bunch of tasks
- Drag 'em up and down or use push-to-top
- Some re-orderings will work correctly
- Some re-orderings will immediately be reversed, or a different task will jump into a different place

**Comment by gisikw** [comment]
Aha! Okay, I got it to repro. Just had to add a bunch of tasks and everything went weird :)

Lemme get a test case in place, and then I should be able to get it fixed.

**Comment by Alys** [comment]
"everything went weird" - yes, that's the bug. That describes it exactly. :-/

**Comment by gisikw** [comment]
Hahahahah.

Yeah, because `nil` is being passed in as the comparison operator, Angular is trying to just sort by comparing the objects themselves, which is a totally unpredictable and stupid thing.

**Comment by gisikw** [comment]
Okay, unfortunately I'm not able to write a failing test case for this issue because it requires some end-to-end testing capability that we don't currently have. But I can at least write a fix, and a test for the fix!

**Comment by crookedneighbor** [comment]
Thank you for being willing, I was thinking the same thing that some work needs to be done on the e2e testing before a test is even possible.

**Comment by gisikw** [comment]
@crookedneighbor Selenium and Firefox seem to be fierce foes.

Any thoughts about setting up Travis integration with SauceLabs, and running e2e with Chrome/Chromium? It unfortunately doesn't help folks run the e2e stuff on vagrant, but it would at least allow some progress on that front.

(Or, ya know, Angular could learn to play nice with PhantomJS)

(Or...we could rewrite this as a React app!!!!) :D

**Comment by gisikw** [comment]
Fixed (and tested) in #5022
